rust/datafusion/src/execution/physical_plan/math_expressions.rs

```diff
@@ -0,0 +1,78 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Math expressions
+
+use crate::error::ExecutionError;
+use crate::execution::context::ExecutionContext;
+use crate::execution::physical_plan::udf::ScalarFunction;
+
+use arrow::array::{Array, ArrayRef, Float64Array, Float64Builder};
+use arrow::datatypes::{DataType, Field};
+
+use std::sync::Arc;
+
+macro_rules! math_unary_function {
+    ($NAME:expr, $FUNC:ident) => {
+        ScalarFunction::new(
+            $NAME,
+            vec![Field::new("n", DataType::Float64, true)],
+            DataType::Float64,
```

Comment:
This looks great, I'll try and do a proper review soon.  Are there any considerations we need to make here for the fact that `math_unary_function` assumes `f64`, will this panic on `f32` input?

Reply:
That's a great point. We need the type coercion optimizer rule to take care of this by automatically casting expressions to the required type where possible or failing at that stage if types are not compatible. I will work on this today.